src/backend/distributed/commands/utility_hook.c

```diff
@@ -1025,31 +1026,55 @@ ExecuteDistributedDDLJob(DDLJob *ddlJob)
 		/*
 		 * Start a new transaction to make sure CONCURRENTLY commands
 		 * on localhost do not block waiting for this transaction to finish.
+		 *
+		 * In addition to doing that, we also need to tell other backends
+		 * --including the ones spawned for connections opened to localhost to
+		 * build indexes on shards of this relation-- that concurrent index
+		 * builds can safely ignore us.
+		 *
+		 * Normally, DefineIndex() only does that if index doesn't have any
+		 * predicates (i.e.: where clause) and no index expressions at all.
+		 * However, now that we already called standard process utility,
+		 * index build on the shell table is finished anyway.
+		 *
+		 * The reason behind doing so is that we cannot guarantee not
+		 * grabbing any snapshots via adaptive executor, and the backends
+		 * creating indexes on local shards (if any) might block on waiting
+		 * for current xact of the current backend to finish, which would
+		 * cause self deadlocks that are not detectable.
 		 */
 		if (ddlJob->startNewTransaction)
 		{
 			/*
-			 * If cache is not populated, system catalog lookups will cause
-			 * the xmin of current backend to change. Then the last phase
-			 * of CREATE INDEX CONCURRENTLY, which is in a separate backend,
-			 * will hang waiting for our backend and result in a deadlock.
-			 *
-			 * We populate the cache before starting the next transaction to
-			 * avoid this. Most of the metadata has already been resolved in
-			 * planning phase, we only need to lookup metadata needed for
-			 * connection establishment.
+			 * To ensure not leaking any snapshots, pop active snapshot if we
+			 * had any.
 			 */
-			(void) CurrentDatabaseName();
+			if (ActiveSnapshotSet())
+			{
```

Comment:
can you elaborate on why/when we can have or not have a snapshot here?

Reply:
Why did we not need this in the PG13 case?

Reply:
You're right, seems that we need to do so for pg<14 too 👍
Let me just copy paste the same section to `#if PG_VERSION_NUM < 140000` block too.

(Could also move to a function such as `PopActiveSnapshotIfExists` but such a function feels slightly confusing & unnecessary to me)

Reply:
Umm, I guess you're actually asking whether we really need to pop the active snapshot or not, let me think a bit more

Reply:
In `DefineIndex`, each time before starting a new xact, postgres seems to be popping the active snapshot, so I thought that we need to do so as well.

However, I guess we might not need to unregister the snapshot since postgres only does that for the TransactionSnapshot that it pushes within `DefineIndex` function itself, not for the snapshot that has been pushed beforehand.